text/0000-explicit-tail-calls.md

```diff
@@ -0,0 +1,774 @@
+<!-- This is the technical portion of the RFC. Explain the design in sufficient detail that:
+
+- Its interaction with other features is clear.
+- It is reasonably clear how the feature would be implemented.
+- Corner cases are dissected by example.
+
+The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work. -->
+Implementation of this feature requires checks that all prerequisites to guarantee TCE are fulfilled.
+These checks are:
+
+- The `become` keyword is only used in place of `return`. The intent is to reuse the semantics of a `return` signifying "the end of a function". See the section on [tail-call-elimination](#tail-call-elimination) for examples.
+- The argument to `become` is a function (or method) call, that exactly matches the function signature and calling convention of the callee. The intent is to ensure a matching ABI. Note that lifetimes may differ as long as they pass borrow checking, see [below](#return-type-coercion) for specifics on the return type.
+- The stack frame of the calling function is reused, this also implies that the function is never returned to. The required checks to ensure this is possible are: no borrows of local variables are passed to the called function (passing local variables by copy/move is ok since that doesn't require the local variable to continue existing after the call), and no further cleanup is necessary. These checks can be done by using the borrow checker as already described in the [section](#difference) showing the difference between `return` and `become` above.
+
+If any of these checks fail a compiler error is issued.
+
+One additional check must be done, if the backend cannot guarantee that TCE will be performed an ICE is issued. It is also suggested to ensure that the invariants provided by the pre-requisites are maintained during compilation, raising an ICE if this is not the case.
+
+The type of the expression `become <call>` is `!` (the never type, see [here](https://doc.rust-lang.org/std/primitive.never.html)). This is consistent with other control flow constructs such as `return`, which also have the type of `!`.
+
+Note that as `become` is a keyword reserved for exactly the use-case described in this RFC there is no backward-compatibility break. This RFC only specifies the use of `become` inside of functions and instead leaves usage outside of functions unspecfied for use by other features.
+
+This feature will have interactions with other features that depend on stack frames, for example, debugging and backtraces. See [drawbacks](#drawbacks) for further discussion.
+
+See below for specifics on interations with other features.
+
+## Mismatches in Mutability
+
+Mismatches in mutability (like `&T` <-> `&mut T`) for arguments and return type of the function signatures are **not** supported. This support requires a guarantee that mutability has no effect on ABI.
```

Comment:
Not sure if this paragraph is needed — `&T` and `&mut T` are different types, so there shouldn't be any questions regarding this. Maybe as a future possibility it would make more sense. Anyway if we'll keep it somewhere, "This support" should be changed to "Supporting this".

Reply:
updated here: https://github.com/rust-lang/rfcs/pull/3407/commits/b7c69e4a21a5d9552a744fb02bcbab16d455b351

While unlikely to happen, I think having a note on this is still worthwhile.